src/input_parsers/OnnxParser.cpp

```diff
@@ -0,0 +1,1265 @@
+    {
+        String inputName = input.name();
+        _shapeMap.insert(inputName, shapeOfInput(input));
+
+        // If this is one of the specified inputs, create new variables
+        if (_inputNames.exists(inputName))
+        {
+            _processedNodes.insert(inputName);
+            _numberOfFoundInputs += 1;
+            Vector<Variable> nodeVars = makeNodeVariables(inputName, true);
+        }
+    }
+
+    // Add shapes for constants
+    for (auto constant : _network.initializer())
+    {
+        String constantName = constant.name();
+        _shapeMap.insert(constantName, shapeOfConstant(constant));
+        _processedNodes.insert(constantName);
+    }
+}
+
+void OnnxParser::validateAllInputsAndOutputsFound()
+{
+    // By this point, all input variables need to have been found
+    if (_numberOfFoundInputs != _inputNames.size())
+    {
+        String errorMessage = "These input variables could not be found:";
+        for (String inputName : _inputNames)
+        {
+            if (!_varMap.exists(inputName))
+            {
+                String space = " ";
+                errorMessage += space + inputName;
+            }
+        }
+        throw MarabouError(MarabouError::ONNX_PARSE_ERROR, errorMessage.ascii());
+    }
+
+    // Mark the output variables
+    for (Variable var : _varMap[_outputName])
+    {
+        _outputVars.append(var);
+    }
+}
+
+void OnnxParser::processGraph(Set<String> &inputNames, String &outputName, InputQuery& query)
+{
+    _inputNames = inputNames;
+    _outputName = outputName;
+
+    initialiseShapeMap();
+    processNode(outputName, true);
+    validateAllInputsAndOutputsFound();
+    getMarabouQuery(query);
+}
+
+/**
+ * @brief Recursively processes the structure of the graph, adding the relevant
+ * constraints for each node as it goes. Takes in a string rather than a NodeProto
+ * as we initially pass in the output which is *not* a NodeProto.
+ *
+ * @param node
+ * @param makeEquations
+ */
+void OnnxParser::processNode(String &nodeName, bool makeEquations)
+{
+    if (_processedNodes.exists(nodeName))
+    {
+        return;
+    }
+
+    if (_inputNames.exists(nodeName))
+    {
+        _numberOfFoundInputs += 1;
+        // If an inputName is an intermediate layer of the network, we don't need to create Marabou
+        // equations for its inputs. However, we still need to call makeMarabouEquations in order to
+        // compute shapes. We just need to set the makeEquations flag to false
+        makeEquations = false;
+    }
+
+    _processedNodes.insert(nodeName);
+
+    List<onnx::NodeProto> nodes = getNodesWithOutput(nodeName);
+    assert(nodes.size() == 1);
+    onnx::NodeProto& node = nodes.front();
+
+    // First recursively process the input nodes.
+    // This ensures that shapes and values of a node's inputs have been computed first.
+    for (auto inputNode : getInputsToNode(node))
+    {
+        processNode(inputNode, makeEquations);
```

Comment:
Does this need to be done in a recursive way? There will be many redundant calls to already processed inputs compared to an iterative inputs-to-outputs traversal of the graph which touches each node exactly once.

Reply:
To be honest I'm not sure, I think it depends on what constraints are on the topology of an ONNX graph? This is a straightforward port of what the existing python code does. 

I don't think the performance hit is worth worrying about though. Each revisited node does a constant time lookup in the `_processedNodes` set and is then is discarded when found.

Reply:
Also how many nodes are actually in the average graph? 50 would be large right? I'm not imagining more than 200 for even the huge NLP networks?

Reply:
From a C++ perspective, every function call is a new stack frame that can be avoided in the reverse traversal, especially since its made only for a map lookout. For example the VNN-Comp uses 5minutes timeout per benchmark, any time not spent parsing is time that can be spent solving the benchmarks. 

So a fully connected network of say 50 nodes per layer will make 50x50=2500 calls recursive calls per layer only to perform a map look up, and 2450 of those will be spurious (since the first node in the layer will already recursively pre-process all the inputs). For 200 nodes per layer that would be 40000 - 200 spurious calls per layer. In any case, this is probably good enough for now. The recursive to iterative solution is common pattern and simple enough to be done after the fact.

Reply:
Oh, got it. I spent too much time with explicit representations for my own good it seems. Yeah, no issues then! Thanks for the clarification.